lib/io/can.cpp

```diff
@@ -9,113 +9,113 @@
     return std::nullopt;
   }
   const int num_bytes_read = read(socket_, &received_message, sizeof(CanFrame));
   if (num_bytes_read < sizeof(CanFrame)) {
-    log_.log(core::LogLevel::kFatal, "Failed to receive CAN data");
+    logger_.log(core::LogLevel::kFatal, "Failed to receive CAN data");
     return std::nullopt;
   }
-  // TODOLater make log_ more elegant
-  log_.log(core::LogLevel::kDebug,
-           "CAN data received, ID:%i, DATA: %i %i %i %i %i %i %i %i",
-           static_cast<int>(received_message.can_id),
-           static_cast<int>(received_message.data[0]),
-           static_cast<int>(received_message.data[1]),
-           static_cast<int>(received_message.data[2]),
-           static_cast<int>(received_message.data[3]),
-           static_cast<int>(received_message.data[4]),
-           static_cast<int>(received_message.data[5]),
-           static_cast<int>(received_message.data[6]),
-           static_cast<int>(received_message.data[7]));
+  // TODOLater make logger more elegant
+  logger_.log(core::LogLevel::kDebug,
+              "CAN message received, ID:%i, DATA: %i %i %i %i %i %i %i %i",
+              static_cast<int>(received_message.can_id),
+              static_cast<int>(received_message.data[0]),
+              static_cast<int>(received_message.data[1]),
+              static_cast<int>(received_message.data[2]),
+              static_cast<int>(received_message.data[3]),
+              static_cast<int>(received_message.data[4]),
+              static_cast<int>(received_message.data[5]),
+              static_cast<int>(received_message.data[6]),
+              static_cast<int>(received_message.data[7]));
 
   return received_message;
 }
 
-CanResult Can::listen()
+io::CanResult Can::listen()
 {
   if (ioctl(socket_, FIONREAD) < sizeof(CanFrame)) { return io::CanResult::kFailure; }
   const auto data = receive();
   if (!data) { return io::CanResult::kFailure; }
-  CanFrame message = data.value();
-  if (!processors_.contains(message.can_id)) {
-    log_.log(core::LogLevel::kInfo, "No CanProccessor associated with id %i", message.can_id);
+  CanFrame message                 = data.value();
+  const auto subscribed_processors = processors_.find(message.can_id);
+  if (subscribed_processors == processors_.end()) {
+    logger_.log(core::LogLevel::kInfo, "No CanProccessor associated with id %i", message.can_id);
     return io::CanResult::kFailure;
   }
-  std::vector<std::shared_ptr<ICanProcessor>> &subscribed_processors = processors_[message.can_id];
-  for (auto &processor : subscribed_processors) {
+  for (auto &processor : subscribed_processors->second) {
     processor->processMessage(message);
   }
   return io::CanResult::kSuccess;
 }
 
 void Can::addCanProcessor(const uint16_t id, std::shared_ptr<ICanProcessor> processor)
 {
-  processors_[id].push_back(processor);
-  log_.log(core::LogLevel::kInfo, "Added processor for id %i", id);
+  if (processors_.contains(id)) { processors_[id].push_back(processor); }
```

Comment:
i don't like the square bracket indexing because it hides some complexity. to me it looks like something as simple as pointer arithmetic whereas in fact you are doing a lookup in a hash table.

also you are still doing the lookup twice.

```cpp
const auto id_and_processors = processors_.find(id);
if (id_and_processors == processors_.end()) {
  processors_.emplace(id, std::vector<...>({ processor }));
} else {
  id_and_processors->second.push_back(processor);
}
```